### PR TITLE
Add Workflow Node Identifier

### PR DESCRIPTION
**Pull request by AlanCoding**
##### SUMMARY
https://github.com/ansible/awx/issues/2032

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
 - API

##### AWX VERSION
```
9.2.0
```

##### ADDITIONAL INFORMATION
This is necessary both for the collection to replace the schema stuff, and for import / export. And for some testing stuff too.

The modules are partially auto-generated. The WFJT one has basic functionality, but I need to put in some more manual work for special handling of things, like an organization field. The WFJT node has had very little attention so far. Need examples and stuff.

### CODE REVIEW

**Comment by AlanCoding** [comment]
recheck

File: awx_collection/plugins/modules/tower_workflow_job_template.py [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+    tower_oauthtoken:
+      description:
+        - The Tower OAuth token to use.
+      required: False
+      type: str
+      version_added: "3.7"
+extends_documentation_fragment: awx.awx.auth
+'''
+
+EXAMPLES = '''
```
Comment (beeankha) [inline awx_collection/plugins/modules/tower_workflow_job_template.py:134]: We need to add some examples here, but could do it in a separate PR

File: awx_collection/plugins/modules/tower_workflow_job_template_node.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+      description:
+        - The Tower OAuth token to use.
+      required: False
+      type: str
+      version_added: "3.7"
+extends_documentation_fragment: awx.awx.auth
+'''
+
+EXAMPLES = '''
+'''
```
Comment (beeankha) [inline awx_collection/plugins/modules/tower_workflow_job_template_node.py:181]: Same comment as above re: example addition.

File: awx_collection/plugins/modules/tower_workflow_job_template.py (hunk 1 above)
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_workflow_job_template.py:134]: That's a good point, this needs something filled in.

**Review by beeankha — APPROVED** [review]

**Comment by beeankha** [comment]
recheck

File: awx/main/models/workflow.py [hunk 3]
```diff
@@ -144,6 +146,18 @@ class WorkflowJobTemplateNode(WorkflowNodeBase):
         related_name='workflow_job_template_nodes',
         on_delete=models.CASCADE,
     )
+    identifier = models.CharField(
+        max_length=512,
+        default=uuid4,
+        blank=False,
+        help_text=_(
+            'An identifier for this node that is unique within its workflow. '
+            'It is copied to workflow job nodes corresponding to this node.'),
```
Comment (ryanpetrello) [inline awx/main/models/workflow.py:155]: Might be smart to go ahead and `db_index=True` this field?

Comment (ryanpetrello) [reply awx/main/models/workflow.py:155]: Also, do you want to put a unique constraint on this field so people don't provide duplicate values?

Comment (AlanCoding) [reply awx/main/models/workflow.py:155]: Questions about db index...

 - Shouldn't it be indexed _together with_ the workflow (job or template)? It's not globally unique, and if you're doing a global filter, you're pretty much doing it wrong. Is indexing together more memory efficient than just making the field indexed?
 - index the WFJT node identifier, or the workflow job identifier, or both?

Also, is `name` not indexed?

https://github.com/ansible/awx/blob/d2998b42f169cec82e0d7a8d2890739237107ced/awx/main/models/base.py#L357-L360

that doesn't sound right

**Comment by AlanCoding** [comment]
FYI, I want to hold off on merge for https://github.com/ansible/awx/pull/6295 to get in first, assuming that can be a near-term merge, which I think it can be.

File: awx/main/models/workflow.py (hunk 3 above)
Comment (ryanpetrello) [reply awx/main/models/workflow.py:155]: > Also, is name not indexed?

I don't think we index names.

Comment (ryanpetrello) [reply awx/main/models/workflow.py:155]: > Shouldn't it be indexed together with the workflow (job or template)?

We could if we wanted to 🤷‍♂.  We just need _some_ index on this, given its intended use.

Comment (AlanCoding) [reply awx/main/models/workflow.py:155]: yeah I don't think this should be a problem at all.

This will be adding an `index_together` in the exact same way as the `unique_together` for the WFJT nodes. for workflow job nodes, we would be adding `index_together`, where no unique constraint exists. There is _some_ concern about the time that will be incurred during the migration adding this. All pre-migration job nodes will have a blank identifier (timing impact unclear).

Also, both of these options I'm talking about are deprecated

https://docs.djangoproject.com/en/3.0/ref/models/options/

Given the small scope of the change here, it makes sense to take the time to move to the modern patterns, and I'll try to do that when I get back around to this branch.

File: awx_collection/plugins/modules/tower_workflow_template.py [hunk 4]
```diff
@@ -143,6 +145,12 @@ def main():
         supports_check_mode=False
     )
 
+    module.deprecate(msg=(
+        "This module is replaced by the combination of tower_workflow_job_template and "
+        "tower_workflow_job_template_node. This uses the old tower-cli and wll be "
+        "removed in 2022."
+    ), version='4.2.0')
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_workflow_template.py:153]: @gundalow this was discussed in the `#ansible-devel` IRC channel.

This is the most compliant that I can make it, adhering to the old 4-release policy.

We plan to have 2 version streams, one for upstream and one for downstream, in both cases corresponding to server version, and we will only reference downstream versions. Anyway, I can't predict version numbers 2 years from now with certainty, so it's a wild guess so that I can pass sanity tests.

@wenottingham might also want to know what I'm trying to do. This is what I think the final form for all our deprecations should be when we inaugurate the downstream.

Comment (wenottingham) [reply awx_collection/plugins/modules/tower_workflow_template.py:153]: If we have our own separate collection, why do we need to follow the prior ansible deprecation policy?

Comment (wenottingham) [reply awx_collection/plugins/modules/tower_workflow_template.py:153]: I would just say "in a future release"

Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_workflow_template.py:153]: > If we have our own separate collection, why do we need to follow the prior ansible deprecation policy?

Our content is going in the ACD. When people upgrade Ansible 2.9 to 2.10, _this_ content is what they'll be running if their playbooks use the modules here.

And I can't say "in a future release". Tests will fail if I don't give it a version. These tests are explicitly designed to enforce standards among collections. I'm not interested in fighting that battle, and frankly, I'm not eager to the action this statement is warning about.

File: awx/main/migrations/0112_v370_workflow_node_identifier.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        ),
+        migrations.RunPython(create_uuid, migrations.RunPython.noop),  # this fixes the uuid4 issue
+        migrations.AlterField(
+            model_name='workflowjobtemplatenode',
+            name='identifier',
+            field=models.CharField(default=uuid.uuid4, help_text='An identifier for this node that is unique within its workflow. It is copied to workflow job nodes corresponding to this node.', max_length=512),
+        ),
+        migrations.AlterUniqueTogether(
+            name='workflowjobtemplatenode',
+            unique_together={('identifier', 'workflow_job_template')},
```
Comment (chrismeyersfsu) [inline awx/main/migrations/0112_v370_workflow_node_identifier.py:47]: `workflowjobnode` also needs an index like this.

File: awx/main/tests/unit/models/test_workflow_unit.py [hunk 6]
```diff
@@ -192,7 +193,8 @@ def test_create_with_prompts(self, wfjt_node_with_prompts, workflow_job_unit, cr
                 char_prompts=wfjt_node_with_prompts.char_prompts,
                 inventory=wfjt_node_with_prompts.inventory,
                 unified_job_template=wfjt_node_with_prompts.unified_job_template,
-                workflow_job=workflow_job_unit)
+                workflow_job=workflow_job_unit,
+                identifier=mocker.ANY)
```
Comment (chrismeyersfsu) [inline awx/main/tests/unit/models/test_workflow_unit.py:197]: TIL

File: awx/main/utils/named_url_graph.py [hunk 7]
```diff
@@ -77,6 +77,8 @@ def _encode_uri(self, text):
         Performance assured: http://stackoverflow.com/a/27086669
         '''
         for c in URL_PATH_RESERVED_CHARSET:
+            if not isinstance(text, str):
+                text = str(text)  # needed for WFJT node creation, identifier temporarily UUID4 type
```
Comment (chrismeyersfsu) [inline awx/main/utils/named_url_graph.py:81]: Can the identifier be a :dragon: ?

File: awx_collection/test/awx/test_workflow_job_template_node.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        'unified_job_template': 'foo-jt',
+        'state': 'present'
+    }, admin_user)
+    assert not result.get('failed', False), result.get('msg', result)
+
+    node = WorkflowJobTemplateNode.objects.get(identifier='42')
+
+    result.pop('invocation', None)
+    assert result == {
+        "name": "42",  # NOTE: should this be identifier instead?
```
Comment (chrismeyersfsu) [inline awx_collection/test/awx/test_workflow_job_template_node.py]: yes

File: awx_collection/test/awx/test_workflow_job_template_node.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+    jt = JobTemplate.objects.create(
+        project=project,
+        inventory=inventory,
+        playbook='helloworld.yml',
+        name='foo-jt'
+    )
+
+    result = run_module('tower_workflow_job_template_node', {
+        'identifier': '42',
```
Comment (chrismeyersfsu) [inline awx_collection/test/awx/test_workflow_job_template_node.py]: splash in some unicode

Comment (AlanCoding) [reply awx_collection/test/awx/test_workflow_job_template_node.py]: I'm not 100% sure the Ansible sanity tests will allow me, but if not, I should be able to add a similar test to our awx functional tests.

**Comment by AlanCoding** [comment]
After applying the last commit I pushed:

```
bash-4.4$ awx-manage dbshell
psql (10.6, server 10.11 (Debian 10.11-1.pgdg90+1))
Type "help" for help.

awx-dev=> select indexname, indexdef from pg_indexes where tablename='main_workflowjobtemplatenode';
                            indexname                            |                                                                                   indexdef                        
                                                            
-----------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------
 main_workfl_identif_0cc025_idx                                  | CREATE INDEX main_workfl_identif_0cc025_idx ON public.main_workflowjobtemplatenode USING btree (identifier)
 main_workfl_identif_6fda75_idx                                  | CREATE INDEX main_workfl_identif_6fda75_idx ON public.main_workflowjobtemplatenode USING btree (identifier, workfl
ow_job_template_id)
 main_workflowjobtemplate_identifier_workflow_job__03484516_uniq | CREATE UNIQUE INDEX main_workflowjobtemplate_identifier_workflow_job__03484516_uniq ON public.main_workflowjobtemp
latenode USING btree (identifier, workflow_job_template_id)
 main_workflowjobtemplatenode_inventory_id_2fab864f              | CREATE INDEX main_workflowjobtemplatenode_inventory_id_2fab864f ON public.main_workflowjobtemplatenode USING btree
 (inventory_id)
 main_workflowjobtemplatenode_workflow_job_template_id_2fd591f0  | CREATE INDEX main_workflowjobtemplatenode_workflow_job_template_id_2fd591f0 ON public.main_workflowjobtemplatenode
 USING btree (workflow_job_template_id)
 main_workflowjobtemplatenode_unified_job_template_id_98b53e6c   | CREATE INDEX main_workflowjobtemplatenode_unified_job_template_id_98b53e6c ON public.main_workflowjobtemplatenode 
USING btree (unified_job_template_id)
 main_workflowjobtemplatenode_pkey                               | CREATE UNIQUE INDEX main_workflowjobtemplatenode_pkey ON public.main_workflowjobtemplatenode USING btree (id)
(7 rows)

awx-dev=> 
awx-dev=> select indexname, indexdef from pg_indexes where tablename='main_workflowjobnode';
                       indexname                       |                                                                indexdef                                                     
            
-------------------------------------------------------+-----------------------------------------------------------------------------------------------------------------------------
------------
 main_workfl_identif_efdfe8_idx                        | CREATE INDEX main_workfl_identif_efdfe8_idx ON public.main_workflowjobnode USING btree (identifier)
 main_workfl_identif_87b752_idx                        | CREATE INDEX main_workfl_identif_87b752_idx ON public.main_workflowjobnode USING btree (identifier, workflow_job_id)
 main_workflowjobnode_inventory_id_1dac2da9            | CREATE INDEX main_workflowjobnode_inventory_id_1dac2da9 ON public.main_workflowjobnode USING btree (inventory_id)
 main_workflowjobnode_workflow_job_id_dcd715c7         | CREATE INDEX main_workflowjobnode_workflow_job_id_dcd715c7 ON public.main_workflowjobnode USING btree (workflow_job_id)
 main_workflowjobnode_unified_job_template_id_8a30f93e | CREATE INDEX main_workflowjobnode_unified_job_template_id_8a30f93e ON public.main_workflowjobnode USING btree (unified_job_t
emplate_id)
 main_workflowjobnode_job_id_key                       | CREATE UNIQUE INDEX main_workflowjobnode_job_id_key ON public.main_workflowjobnode USING btree (job_id)
 main_workflowjobnode_pkey                             | CREATE UNIQUE INDEX main_workflowjobnode_pkey ON public.main_workflowjobnode USING btree (id)
(7 rows)

```

before 5 indexes on both models
after, 7 indexes on both models

Existing indexes not modified, only added 2 for the `identifier` field and `('identifier', <parent workflow thing>)`

File: awx_collection/plugins/module_utils/tower_api.py [hunk 10]
```diff
@@ -603,7 +612,11 @@ def update_if_needed(self, existing_item, new_item, on_update=None, associations
 
         # If we change something and have an on_change call it
         if on_update is not None and self.json_output['changed']:
-            on_update(self, response['json'])
+            if response is None:
+                last_data = existing_item
+            else:
+                last_data = response['json']
+            on_update(self, last_data)
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:619]: @beeankha @john-westcott-iv I do want to avoid changing module_utils whenever possible

but I just rebased and this still accounts for a bug

```
===================================================================================== FAILURES ======================================================================================
___________________________________________________________________________ test_survey_spec_only_changed ___________________________________________________________________________
Traceback (most recent call last):
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/test_workflow_job_template.py", line 58, in test_survey_spec_only_changed
    }, admin_user)
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/conftest.py", line 146, in rf
    resource_module.main()
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/modules/tower_workflow_job_template.py", line 234, in main
    on_create=on_change, on_update=on_change
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/module_utils/tower_api.py", line 626, in create_or_update_if_needed
    return self.update_if_needed(existing_item, new_item, on_update=on_update, associations=associations)
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/module_utils/tower_api.py", line 615, in update_if_needed
    on_update(self, response['json'])
TypeError: 'NoneType' object is not subscriptable
------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------

```

that's fixed with this change, may just be the first to hit it

File: awx_collection/plugins/module_utils/tower_api.py [hunk 11]
```diff
@@ -556,6 +562,7 @@ def update_if_needed(self, existing_item, new_item, on_update=None, associations
         #    2. The response from Tower from patching to the endpoint. It's up to you to process the response and exit from the module.
         #    3. An ItemNotDefined exception, if the existing_item does not exist
         # Note: common error codes from the Tower API can cause the module to fail
+        response = None
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:565]: this corresponds to

```
===================================================================================== FAILURES ======================================================================================
___________________________________________________________________________ test_survey_spec_only_changed ___________________________________________________________________________
Traceback (most recent call last):
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/test_workflow_job_template.py", line 58, in test_survey_spec_only_changed
    }, admin_user)
  File "/Users/alancoding/Documents/tower/awx_collection/test/awx/conftest.py", line 146, in rf
    resource_module.main()
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/modules/tower_workflow_job_template.py", line 234, in main
    on_create=on_change, on_update=on_change
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/module_utils/tower_api.py", line 624, in create_or_update_if_needed
    return self.update_if_needed(existing_item, new_item, on_update=on_update, associations=associations)
  File "/Users/alancoding/Documents/tower/awx_collection/plugins/module_utils/tower_api.py", line 614, in update_if_needed
    if response is None:
UnboundLocalError: local variable 'response' referenced before assignment
------------------------------------------------------------------------------- Captured stdout call --------------------------------------------------------------------------------
```

**Comment by AlanCoding** [comment]
recheck

**Comment by AlanCoding** [comment]
Here's my demo playbook content for this:

https://gist.github.com/AlanCoding/e478a62fd77f2ceae507a553e31dc5e9

**Comment by AlanCoding** [comment]
I did the indexes wrong before. The WFJT node indexes had 2 entries for (identifier, wfjt_id). This is what it is now.

```
awx-dev=> select indexname, indexdef from pg_indexes where tablename='main_workflowjobtemplatenode';
                            indexname                            |                                                                                   indexdef                        
                                                            
-----------------------------------------------------------------+-------------------------------------------------------------------------------------------------------------------
------------------------------------------------------------
 main_workfl_identif_0cc025_idx                                  | CREATE INDEX main_workfl_identif_0cc025_idx ON public.main_workflowjobtemplatenode USING btree (identifier)
 main_workflowjobtemplate_identifier_workflow_job__03484516_uniq | CREATE UNIQUE INDEX main_workflowjobtemplate_identifier_workflow_job__03484516_uniq ON public.main_workflowjobtemp
latenode USING btree (identifier, workflow_job_template_id)
 main_workflowjobtemplatenode_inventory_id_2fab864f              | CREATE INDEX main_workflowjobtemplatenode_inventory_id_2fab864f ON public.main_workflowjobtemplatenode USING btree
 (inventory_id)
 main_workflowjobtemplatenode_workflow_job_template_id_2fd591f0  | CREATE INDEX main_workflowjobtemplatenode_workflow_job_template_id_2fd591f0 ON public.main_workflowjobtemplatenode
 USING btree (workflow_job_template_id)
 main_workflowjobtemplatenode_unified_job_template_id_98b53e6c   | CREATE INDEX main_workflowjobtemplatenode_unified_job_template_id_98b53e6c ON public.main_workflowjobtemplatenode 
USING btree (unified_job_template_id)
 main_workflowjobtemplatenode_pkey                               | CREATE UNIQUE INDEX main_workflowjobtemplatenode_pkey ON public.main_workflowjobtemplatenode USING btree (id)
(6 rows)
```

Because the unique_together automatically adds an index.
